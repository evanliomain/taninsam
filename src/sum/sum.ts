/**
 * @module array=>number
 */
/**
 * This method sum all element of the chained array.
 * @return the function to apply on the array
 * to sum its elements and return the sum
 * @example
 * ```
 * sum()([1, 2, 3, 4, 5]) // 15
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(sum())
 *   .value() // 15
 * ```
 */
export function sum(): (elements: ReadonlyArray<number>) => number {
  return (elements: ReadonlyArray<number>) =>
    elements.reduce((s, element) => s + element, 0);
}
