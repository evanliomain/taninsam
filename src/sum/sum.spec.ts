import { describe, expect, test } from 'vitest';
import { sum } from './sum';

describe('sum function', () => {
  const input = [1, 2, 3, 4, 5, 6, 7];

  test('is a pure function', () => {
    sum()(input);
    expect(input).toMatchSnapshot();
  });

  test('input |> sum === 28', () => {
    expect(sum()(input)).toBe(28);
  });

  test('[] |> sum === 0', () => {
    expect(sum()([])).toBe(0);
  });
});
