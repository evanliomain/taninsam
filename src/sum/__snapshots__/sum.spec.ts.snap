// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sum function > is a pure function 1`] = `
[
  1,
  2,
  3,
  4,
  5,
  6,
  7,
]
`;
